### PR TITLE
Credit transaction fees to the slot leader

### LINKED ISSUE #2623: Transaction fee isn't collected by the slot leader
The fee for a Transaction is currently debited
https://github.com/solana-labs/solana/blob/1c0758e3bdf4fb897d063294aec9793e947db1da/src/accounts.rs#L152
but it is not credited anywhere, effectively burning the fee.

Instead the Transaction fee should be credited to the slot leader's staking account.

cc: #1527

### PR DESCRIPTION
**Pull request by pgarg66**
#### Problem
The transaction fees is not credited to the slot leader

#### Summary of Changes
Calculate the transaction fee for the leader slot duration, and credit it to the leader's account at slot boundary

Fixes #2623

### CODE REVIEW

**Comment by pgarg66** [comment]
Could you review the overall flow of the code change? I want to make sure I am approaching it from the right direction.

The ledger processing needs to also credit the slot leader. I'll plumb that path if things look ok here.

**Comment by garious** [comment]
I was thinking of something a little more straightforward. In the ReplayStage, run:

```rust
let fees = entries.iter().map(|entry| entry.transactions.map(|tx| tx.fee).sum()).sum();
bank.transfer(fees, leader_id);
```

**Comment by garious** [comment]
but of course not using `transfer()`. Using `slow_store` as you're doing

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -534,6 +537,13 @@ impl Bank {
         parents
     }
 
+    pub fn collect_tx_fee(&self, pubkey: &Pubkey, fee: u64) {
```
Comment (garious) [inline runtime/src/bank.rs]: You can call that `deposit`

**Comment by garious** [comment]
On closer inspection, that's looking about right. I'm surprised you needed to dip all the way into runtime though.  Didn't have enough info in `Bank::commit_transactions()`?

**Comment by garious** [comment]
Just please don't plumb the leader_id into any `solana_runtime` modules. I hope that logic can be managed at a higher layer?

**Comment by pgarg66** [comment]
> Just please don't plumb the leader_id into any `solana_runtime` modules. I hope that logic can be managed at a higher layer?

So, in current patch, runtime/src/bank.rs has a new function collect_tx_fee (renaming it to deposit). It gets a pubkey as a parameter. Is that ok?

**Comment by garious** [comment]
> So, in current patch, runtime/src/bank.rs has a new function collect_tx_fee (renaming it to deposit). It gets a pubkey as a parameter. Is that ok?

Yes, that's perfect

File: src/blocktree_processor.rs [hunk 2]
```diff
@@ -53,22 +53,27 @@ fn ignore_program_errors(results: Vec<Result<()>>) -> Vec<Result<()>> {
...
             bank.unlock_accounts(&e.transactions, &results);
-            first_err(&results)
+            (first_err(&results), fee)
         })
         .collect();
-    first_err(&results)
+
+    let fee = results_fees.iter().map(|(_, fee)| fee).sum();
+    let mut results = Vec::new();
+    results.reserve_exact(results_fees.len());
```
Comment (garious) [inline src/blocktree_processor.rs]: You shouldn't need this or the mut. map() or collect() should know to do this

File: src/replay_stage.rs [hunk 3]
```diff
@@ -123,6 +127,19 @@ impl ReplayStage {
 
                 if 0 == num_ticks_to_next_vote {
                     subscriptions.notify_subscribers(&bank);
+                    let slot_num = leader_scheduler
```
[the remark this answers is not in the document]
Comment (pgarg66) [reply src/replay_stage.rs]: 👍

File: src/blocktree_processor.rs (hunk 2 above)
Comment (pgarg66) [reply src/blocktree_processor.rs]: 👍

**Comment by pgarg66** [comment]
I'll look into adding some tests

**Comment by pgarg66** [comment]
> On closer inspection, that's looking about right. I'm surprised you needed to dip all the way into runtime though. Didn't have enough info in `Bank::commit_transactions()`?

I was able to remove most of the changes from runtime.

File: runtime/src/bank.rs [hunk 4]
```diff
@@ -534,6 +534,13 @@ impl Bank {
         parents
     }
 
+    pub fn deposit(&self, pubkey: &Pubkey, fee: u64) {
+        if let Some(mut account) = self.get_account(pubkey) {
```
Comment (garious) [inline runtime/src/bank.rs]: If None, create the account.

File: src/blocktree_processor.rs [hunk 5]
```diff
@@ -132,11 +151,26 @@ fn process_block(
     leader_scheduler: &Arc<RwLock<LeaderScheduler>>,
 ) -> Result<()> {
     for entry in entries {
-        process_entry(bank, entry)?;
+        let (res, fee) = process_entry(bank, entry);
+        let num_ticks = bank.tick_height();
```
Comment (garious) [inline src/blocktree_processor.rs]: `tick_height`

File: src/blocktree_processor.rs [hunk 6]
```diff
@@ -132,11 +151,26 @@ fn process_block(
...
+        let slot_height = leader_scheduler
```
Comment (garious) [inline src/blocktree_processor.rs]: Before merging this, can you add `get_leader_for_tick` method to `LeaderScheduler` and replace all the instances of this exact pattern?

Comment (pgarg66) [reply src/blocktree_processor.rs]: sure

File: src/replay_stage.rs [hunk 7]
```diff
@@ -123,6 +127,19 @@ impl ReplayStage {
 
                 if 0 == num_ticks_to_next_vote {
                     subscriptions.notify_subscribers(&bank);
+                    let slot_height = leader_scheduler
```
Comment (garious) [inline src/replay_stage.rs]: If you don't add that method, some shorter copy pasta for you:

```rust
let leader = {
   let leader_scheduler = leader_scheduler.read().unwrap();
   let slot_height = leader_scheduler.tick_height_to_slot(num_ticks);
   leader_scheduler.get_leader_for_slot(slot_height)
};
```

**Comment by rob-solana** [comment]
doing this above runtime means that the TPU's bank won't match the TVU's, right?

**Pull request merged**